Add snap packaging to .goreleaser.yml

jnsgruk opened the pull request:
This PR adds snap packaging to the goreleaser config.

I've set `publish: false` for now, as we should discuss the release strategy and transfer the `parca` name in the Snap Store to you folks when/if you're ready to take it on?

I've yet to add anything to the CI config, but my proposal is this:

- For "proper" version releases, push a version to `latest/stable`. E.g. `0.12.2` would  go straight into `latest/stable`
- For merges onto `main`, release automatically into `latest/edge`

Let me know if you're happy with that, and I'll update the Github Workflows.

One other option is to use [branches](https://snapcraft.io/blog/time-to-branch-out) on each PR, so that every PR would auto-publish to e.g. `latest/edge/pr-25` . This makes for some pretty easy QA if you're interested ;-)

My only other consideration is on the version template for snapshot releases - we could change to something like:

```yaml
snapshot:
  name_template: "{{ incpatch .Tag }}-{{ .Commit }}"
```

So that people can more easily see which revision is sitting in `latest/edge`?

kakkoyun commented on snap/parca-wrapper:
```diff
@@ -0,0 +1,35 @@
+fi
+
+# Run parca with the gathered arguments
+exec ${SNAP}/parca ${opts[@]}
```
```suggestion
exec ${SNAP}/parca ${opts[@]}

```

kakkoyun reviewed (commented):
Nice!!!!

kakkoyun commented:
> I've yet to add anything to the CI config, but my proposal is this:
> 
> * For "proper" version releases, push a version to `latest/stable`. E.g. `0.12.2` would  go straight into `latest/stable`
> * For merges onto `main`, release automatically into `latest/edge`

This sounds perfect! 

> My only other consideration is on the version template for snapshot releases - we could change to something like:
> 
> ```yaml
> snapshot:
>   name_template: "{{ incpatch .Tag }}-{{ .Commit }}"
> ```

This also sounds good to me. Right now, I think it's something like `Tag-next`

jnsgruk commented:
Alright, sweet. I'll adjust the snapshot name template. Just trying to figure out how to make goreleaser be more discerning about when/how it publishes to which channel.

jnsgruk commented on .goreleaser.yml:
```diff
@@ -31,10 +31,72 @@ archives:
+      Continuous profiling for analysis of CPU and memory usage, down to the line
+      number and throughout time. Saving infrastructure cost, improving performance,
+      and increasing reliability.
+    # This allows control over when the snap is published into `stable`
+    grade: "{{ if .IsSnapshot }}devel{{ else }}stable{{ end }}"
```
I've gone for this approach at the moment, because I can't see an obvious way to control the releases such that snapshots don't end up in `stable`. Got an open question with the goreleaser folk so will revisit once I get an answer.

jnsgruk replied:
This didn't age well :sweat_smile:

maxbrunet reviewed (commented):
Mo' packaging!

Please ensure all shellchecks are addressed:

```console
$ shellcheck snap/hooks/configure

In snap/hooks/configure line 5:
SNAP=$(echo $SNAP | sed -e "s|/var/lib/snapd||g")
       ^-- SC2001 (style): See if you can use ${variable//search/replace} instead.
            ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
SNAP=$(echo "$SNAP" | sed -e "s|/var/lib/snapd||g")


In snap/hooks/configure line 27:
mkdir -p $SNAP_DATA/profiles
         ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
mkdir -p "$SNAP_DATA"/profiles


In snap/hooks/configure line 28:
cp -pnr $SNAP/usr/share/parca/example-config.yaml $SNAP_DATA/parca.yaml
        ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.
                                                  ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
cp -pnr "$SNAP"/usr/share/parca/example-config.yaml "$SNAP_DATA"/parca.yaml

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
  https://www.shellcheck.net/wiki/SC2001 -- See if you can use ${variable//se...
```

```console
$ shellcheck snap/parca-wrapper

In snap/parca-wrapper line 4:
${SNAP}/snap/hooks/configure
^-----^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
"${SNAP}"/snap/hooks/configure


In snap/parca-wrapper line 35:
exec ${SNAP}/parca ${opts[@]}
     ^-----^ SC2086 (info): Double quote to prevent globbing and word splitting.
                   ^--------^ SC2068 (error): Double quote array expansions to avoid re-splitting elements.

Did you mean:
exec "${SNAP}"/parca ${opts[@]}

For more information:
  https://www.shellcheck.net/wiki/SC2068 -- Double quote array expansions to ...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

And if we could make sure, they are formatted (Google-style w/ 4-space indent), it'd be great:

```console
$ shfmt -i 4 -ci -bn -d snap/parca-wrapper
--- snap/parca-wrapper.orig
+++ snap/parca-wrapper
@@ -11,8 +11,8 @@

 # Start building an array of command line options
 opts=(
-    "--config-path=${SNAP_DATA}/parca.yaml"
-    "--log-level=${log_level}"
+    "--config-path=${SNAP_DATA}/parca.yaml"
+    "--log-level=${log_level}"
     "--port=:${port}"
 )
```

The above are actually trailing whitespaces only, so it is close. I'll add a `.editorconfig` to take care of those automatically (and it'll also deal with the EOF Kemal commented on)

There is no CI for all that at moment, but I do not know if they would have been caught since they do not have any file extension. I'll try to look into that

jnsgruk commented:
Sure! I'll take a look this evening, just stepping out for food!

jnsgruk commented on .goreleaser.yml:
```diff
@@ -31,10 +31,71 @@ archives:
+    # * `devel` = ["edge", "beta"]
+    channel_templates:
+      - edge
+      - "{{ if not .IsSnapshot }}stable{{ end }}"
```
This is now the only thing I'm not sure about - I don't know what'll happen if this tries to publish. Still spelunking in goreleaser.

jnsgruk commented:
Ready for another look - I made a terrible rebasing mistake but the diff is clean again :sweat_smile:

kakkoyun commented on .goreleaser.yml:
```diff
@@ -31,10 +31,71 @@ archives:
   name_template: 'checksums.txt'
 snapshot:
-  name_template: "{{ incpatch .Tag }}-next"
+  name_template: "{{ incpatch .Tag }}-{{ .ShortCommit }}"
```
:+1:

kakkoyun reviewed (approved):
LGTM. As you said, we should test it before merging.

jnsgruk commented:
Alright I think we're good! @kakkoyun I've added a new workflow file specifically for building and testing the snap. It will also try to release the snap on merges to main (which will fail until we setup a token in the CI).

I've ended up not using the publish feature of goreleaser for the snap; it's a little bit too primitive and I think we can do better with just the normal snapcraft publishing feature - you'll get more control over which revisions go into which channel.

Behaviour now:

- Every PR will build the snap, install the snap and test the config options
- Merges to `main` will do the same, but also release to `latest/edge`
- Tags starting `v*` will be released to `stable`.

There was a bit of collatoral damage to single quotes, I went back and changed them a few times, but settled on having them all consistently using double quotes - let me know if you'd like me to revert those changes.

A possible improvement, likely for a future PR, would be a bit of a re-org of the workflow so that the snapshot release is only built once per pipeline - saving some time (and the planet) on each run! :) I'm happy to tackle that some day if that's useful!

kakkoyun reviewed (approved):
LGTM!

Thanks for the awesomeness :heart:

simskij commented:
Great work, @jnsgruk! 🎉
